twilight-gateway/src/shard.rs

```diff
@@ -0,0 +1,987 @@
+    ///
+    /// ```no_run
+    /// # #[tokio::main] async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use std::{borrow::Cow, env};
+    /// use twilight_gateway::{
+    ///     message::{CloseFrame, Message},
+    ///     Intents,
+    ///     Shard,
+    ///     ShardId,
+    /// };
+    ///
+    /// let token = env::var("DISCORD_TOKEN")?;
+    /// let mut shard = Shard::new(ShardId::ONE, token, Intents::GUILDS).await?;
+    ///
+    /// let message = Message::Close(Some(CloseFrame::NORMAL));
+    /// shard.send(message).await?;
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`SendErrorType::Sending`] error type if the message could
+    /// not be sent over the websocket. This indicates the shard is either
+    /// currently restarting or closed and will restart.
+    ///
+    /// [ratelimiter]: CommandRatelimiter
+    /// [was enabled]: crate::ConfigBuilder::ratelimit_messages
+    pub async fn send(&mut self, message: Message) -> Result<(), SendError> {
+        if let Some(ref ratelimiter) = self.ratelimiter {
+            ratelimiter.acquire_one().await;
+        }
+
+        self.connection
+            .send(message.into_tungstenite())
+            .await
+            .map_err(|source| {
+                self.disconnect(Disconnect::Resume);
+
+                SendError {
+                    kind: SendErrorType::Sending,
+                    source: Some(Box::new(source)),
+                }
+            })
+    }
+
+    /// Retrieve a channel to send messages over the shard to the gateway.
+    ///
+    /// This is primarily useful for sending to other tasks and threads where
+    /// the shard won't be available.
+    pub fn sender(&mut self) -> MessageSender {
+        self.user_channel.sender()
+    }
+
+    /// Close the shard's connection, providing a close frame indicating whether
+    /// a resume is intended.
+    ///
+    /// Returns the gateway session of the shard, which may be provided via
+    /// [`ConfigBuilder::session`] to create a new shard that will resume the
+    /// gateway session.
+    ///
+    /// If sending a close frame such as [`CloseFrame::NORMAL`] then Discord
+    /// will invalidate the shard's session, showing the application's bot as
+    /// offline. If sending a close frame such as [`CloseFrame::RESUME`] then
+    /// Discord will not invalidate the shard's session and will continue to
+    /// show the application's bot as online until its presence times out.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`SendErrorType::Sending`] error type if the message could
+    /// not be sent over the websocket. This indicates the shard is either
+    /// currently restarting or closed and will restart.
+    ///
+    /// [`ConfigBuilder::session`]: crate::ConfigBuilder::session
+    pub async fn close(
+        &mut self,
+        maybe_close_frame: Option<CloseFrame<'static>>,
```

Comment:
This should either require a `CloseFrame` or clearly document the outcome of passing `None`.

Reply:
I'm not sure what I would document here. If no close frame is provided then no close frame is sent. Is there a different possibility?

Reply:
And what's the outcome of not sending a close frame?